src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.WebAssembly.cs

```diff
@@ -96,6 +101,237 @@ private unsafe bool JsEndsWith(ReadOnlySpan<char> source, ReadOnlySpan<char> pre
+        {
+            Debug.Assert(!GlobalizationMode.Invariant);
+
+            Debug.Assert(!target.IsEmpty);
+            Debug.Assert(_isAsciiEqualityOrdinal && CanUseAsciiOrdinalForOptions(options));
+
+            exceptionMessage = "";
+
+            fixed (char* ap = &MemoryMarshal.GetReference(source))
+            fixed (char* bp = &MemoryMarshal.GetReference(target))
+            {
+                char* a = ap;
+                char* b = bp;
+
+                for (int j = 0; j < target.Length; j++)
+                {
+                    char targetChar = *(b + j);
+                    if (targetChar >= 0x80 || HighCharTable[targetChar])
+                        goto InteropCall;
+                }
+
+                if (target.Length > source.Length)
+                {
+                    for (int k = 0; k < source.Length; k++)
+                    {
+                        char targetChar = *(a + k);
+                        if (targetChar >= 0x80 || HighCharTable[targetChar])
+                            goto InteropCall;
+                    }
+                    return -1;
+                }
+
+                int startIndex, endIndex, jump;
+                if (fromBeginning)
+                {
+                    // Left to right, from zero to last possible index in the source string.
+                    // Incrementing by one after each iteration. Stop condition is last possible index plus 1.
+                    startIndex = 0;
+                    endIndex = source.Length - target.Length + 1;
+                    jump = 1;
+                }
+                else
+                {
+                    // Right to left, from first possible index in the source string to zero.
+                    // Decrementing by one after each iteration. Stop condition is last possible index minus 1.
+                    startIndex = source.Length - target.Length;
+                    endIndex = -1;
+                    jump = -1;
+                }
+
+                for (int i = startIndex; i != endIndex; i += jump)
+                {
+                    int targetIndex = 0;
+                    int sourceIndex = i;
+
+                    for (; targetIndex < target.Length; targetIndex++, sourceIndex++)
+                    {
+                        char valueChar = *(a + sourceIndex);
+                        char targetChar = *(b + targetIndex);
+
+                        if (valueChar >= 0x80 || HighCharTable[valueChar])
+                            goto InteropCall;
+
+                        if (valueChar == targetChar)
+                        {
+                            continue;
+                        }
+
+                        // The match may be affected by special character. Verify that the following character is regular ASCII.
+                        if (sourceIndex < source.Length - 1 && *(a + sourceIndex + 1) >= 0x80)
+                            goto InteropCall;
```

Comment:
> Are we wasting CPU cycles on average by doing the work again in JS ? 

Yes, when the locale is `en-*` and the compariosn option is not `IgnoreSymbols`. 

> Could we pass the current position and continue the search ?

That makes sense. Because everyting up to the current position is equal, it pays of to start comparison in JS from that index, not from the beginning. But continuation of search is not beneficial - we will just compare all the remaining string in the InteropCall because we established that it's not a regular all-ASCII.

Reply:
We cannot do it due to surogates. Example:
source: "o\u0308" = ö
needle: "o"
expected result: -1
If we do: check char by char till you get to some non-ASCII, save from what idx you had the match (from idx=0) and then send **the rest** of the strings (source: "\u0308", needle: "") to continue in JS - we get result = 0.
We need to keep the helper, it's a good optimisation in most cases. Only cases with "lotsOfAsciiAndThenAUnicodeAtTheEndThatWasNotCaughtByTheRegex" will suffer here, so not many. We could get rid of it totally and just send all strings to JS but then the "allAscii" cases would suffer.